Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/features/input/ui/AdvancedMappingDialog.kt

```diff
@@ -0,0 +1,124 @@
+        binding.dropdownDevice.isSaveEnabled = false
+
+        binding.dropdownDevice.onItemClickListener = this
+
+        val deviceAdapter =
+            ArrayAdapter(context, android.R.layout.simple_spinner_dropdown_item, devices!!)
+        binding.dropdownDevice.setAdapter(deviceAdapter)
+
+        controlAdapter =
+            AdvancedMappingControlAdapter { control: String -> onControlClicked(control) }
+        binding.listControl.adapter = controlAdapter
+        binding.listControl.layoutManager = LinearLayoutManager(context)
+
+        val divider = MaterialDividerItemDecoration(context, LinearLayoutManager.VERTICAL)
+        divider.isLastItemDecorated = false
+        binding.listControl.addItemDecoration(divider)
+
+        binding.editExpression.setText(controlReference.getExpression())
+
+        selectDefaultDevice()
+    }
+
+    val expression: String
+        get() = binding.editExpression.text.toString()
+
+    override fun onItemClick(adapterView: AdapterView<*>?, view: View, position: Int, id: Long) =
+        setSelectedDevice(devices?.get(position)!!)
+
+    private fun setSelectedDevice(deviceString: String) {
+        selectedDevice = deviceString
+
+        val device = getDevice(deviceString)
+        if (device == null)
+            setControls(arrayOfNulls(0))
+        else if (controlReference.isInput())
+            setControls(device.getInputs())
+        else
+            setControls(device.getOutputs())
+    }
+
+    private fun setControls(controls: Array<CoreDevice.Control?>) {
+        controlAdapter.setControls(controls.map { it?.getName() }.toTypedArray())
+    }
+
+    private fun onControlClicked(control: String) {
+        val expression =
+            getExpressionForControl(control, selectedDevice!!, controller.getDefaultDevice())
+
+        val start = binding.editExpression.selectionStart.coerceAtLeast(0)
+        val end = binding.editExpression.selectionEnd.coerceAtLeast(0)
+        binding.editExpression.text?.replace(
+            start.coerceAtMost(end),
+            start.coerceAtLeast(end),
+            expression,
+            0,
+            expression.length
+        )
+    }
+
+    private fun selectDefaultDevice() {
+        val defaultDevice = controller.getDefaultDevice()
+        val isInput = controlReference.isInput()
+
+        if (listOf(*devices!!).contains(defaultDevice) &&
+            (isInput || deviceHasOutputs(defaultDevice))
+        ) {
+            // The default device is available, and it's an appropriate choice. Pick it
+            setSelectedDevice(defaultDevice)
+            binding.dropdownDevice.setText(defaultDevice, false)
+            return
+        } else if (!isInput) {
+            // Find the first device that has an output. (Most built-in devices don't have any)
+            val deviceWithOutputs = devices.first { deviceHasOutputs(it!!) }
+            if (deviceWithOutputs!!.isNotEmpty()) {
```

Comment:
Why is there both `!!` and `isNotEmpty` here? I'm a little confused.

Reply:
This isn't the case anymore since I changed `deviceWithOutputs` to be non-nullable but `isNotEmpty` operates on non-null Strings. It really only checks for `""` and not `null`. You could do this another way by turning a `String?` into a `String` by using `toString` which can convert `null` into `""`.